presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotClusterInfoFetcher.java

```diff
@@ -110,6 +116,9 @@ public PinotClusterInfoFetcher(
         this.brokersForTableCache = CacheBuilder.newBuilder()
                 .expireAfterWrite(cacheExpiryMs, TimeUnit.MILLISECONDS)
                 .build((CacheLoader.from(this::getAllBrokersForTable)));
+        this.instanceConfigCache = CacheBuilder.newBuilder()
```

Comment:
Pinot instance is in general stateful and won't ***change*** (much). Restart won't change instance/port etc.
Cache will be expired based on the configurable time.

This cache is to avoid call to pinot controller per request(This is 1 call per split, which means if we have 100 segments and using 1 split per segment, then every query will issue 100 calls to pinot controller for this information)

Reply:
Hmm. Why ? At the very least, I would expect to be making one call per server to get that server's GRPC port. Why is it per split ? I see that the cache is working in lieu of a batched API at the controller.

Reply:
In order to get the grpc port, we need to query pinot controller based on pinot instance name.

When doing segment level query, we issue one query for every split (which contains the pinot server to query and the segments to query). For non-grpc query, the port is part of instance id to be parsed, however this is not true for grpc, we need to query pinot controller for grpc port per pinot instance. 

In the worst case scenario,  assume one pinot table has 100 segments, then one presto query could be split into one segment per split, which means, we will issue 100 queries to pinot servers which means we also need to make 100 calls to pinot controllers for instance grpc port and maybe most of them are asking for the same information.

Hence I want to cache the grpc port here. The first query will go to pinot controller to fetch instance config then extract the grpc port. This info will be cached for 2 minutes. Then the next call will be issued along with the query.

Reply:
Sounds good. This awesome explanation deserves in the code comments :-) !